Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift

```diff
@@ -662,33 +662,21 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
     // Detect C-style for loops based on two semicolons which could not be parsed between the 'for' keyword and the '{'
     // This is mostly a proof-of-concept implementation to produce more complex diagnostics.
     if let unexpectedCondition = node.body.unexpectedBeforeLeftBrace,
-      unexpectedCondition.tokens(withKind: .semicolon).count == 2
+      unexpectedCondition.tokens(withKind: .semicolon).count == 2,
+      let firstToken = node.pattern.firstToken(viewMode: .sourceAccurate),
+      let lastToken = unexpectedCondition.lastToken(viewMode: .sourceAccurate)
     {
-      // FIXME: This is aweful. We should have a way to either get all children between two cursors in a syntax node or highlight a range from one node to another.
       addDiagnostic(
         node,
         .cStyleForLoop,
-        highlights: ([
-          Syntax(node.pattern),
-          Syntax(node.unexpectedBetweenPatternAndTypeAnnotation),
-          Syntax(node.typeAnnotation),
-          Syntax(node.unexpectedBetweenTypeAnnotationAndInKeyword),
-          Syntax(node.inKeyword),
-          Syntax(node.unexpectedBetweenInKeywordAndSequenceExpr),
-          Syntax(node.sequenceExpr),
-          Syntax(node.unexpectedBetweenSequenceExprAndWhereClause),
-          Syntax(node.whereClause),
-          Syntax(node.unexpectedBetweenWhereClauseAndBody),
-          Syntax(unexpectedCondition),
-        ] as [Syntax?]).compactMap({ $0 }),
+        highlights: (getTokens(between: firstToken, and: lastToken).map { Optional(Syntax($0)) }).compactMap({ $0 }),
```

Comment:
Why do you need to explicitly spell out `Optional` here?

Reply:
I was thinking about replicating `as [Syntax?]`,
However, the test failing is `testColoringWithHighlights()`. The problem seems to be that with `getTokens` we are creating an array of tokens, which is not what was happening previously. Do you think I should find a way to walk through `node`'s children?